entrypoint/src/mill/entrypoint/MillBootstrapModule.scala

```diff
@@ -0,0 +1,214 @@
+
+      super.resolveDeps(deps, false)()
+    }
+    def scalaVersion = "2.13.10"
+
+    def parseBuildFiles = T.input {
+      MillBootstrapModule.parseBuildFiles(base)
+    }
+
+    def ivyDeps = T {
+      Agg.from(parseBuildFiles()._2.map(mill.scalalib.Dep.parse(_))) ++
+      Seq(ivy"com.lihaoyi::mill-moduledefs:${Versions.millModuledefsVersion}")
+    }
+
+    def scriptSources = T.sources {
+      for ((p, s) <- parseBuildFiles()._1) yield PathRef(p)
+    }
+
+    def generatedSources = T {
+      val (scripts, ivy, importTrees, errors) = parseBuildFiles()
+      if (errors.nonEmpty) Result.Failure(errors.mkString("\n"))
+      else Result.Success(
+        for ((p, (_, s)) <- scriptSources().zip(parseBuildFiles()._1)) yield {
+          val relative = p.path.relativeTo(base)
+          val segments = Seq.fill(relative.ups)("^") ++ relative.segments
+          val dest = T.dest / segments
+
+          os.write(
+            dest,
+            MillBootstrapModule.top(relative, p.path / os.up, segments.dropRight(1), p.path.baseName) +
+            s +
+            MillBootstrapModule.bottom,
+            createFolders = true
+          )
+          PathRef(dest)
+        }
+      )
+    }
+
+    def scriptImportGraph = T {
+      val (scripts, ivy, scriptImportGraph, errors) = parseBuildFiles()
+      scriptImportGraph
+    }
+
+    override def allSourceFiles: T[Seq[PathRef]] = T {
+      Lib.findSourceFiles(allSources(), Seq("scala", "java", "sc")).map(PathRef(_))
+    }
+
+    def unmanagedClasspath = mill.define.Target.input {
+      mill.api.Loose.Agg.from(enclosingClasspath.map(p => mill.api.PathRef(p))) ++
+      lineNumberPluginClasspath()
+    }
+
+    def scalacPluginIvyDeps = Agg(
+      ivy"com.lihaoyi:::scalac-mill-moduledefs-plugin:${Versions.millModuledefsVersion}"
+    )
+
+    def scalacPluginClasspath = super.scalacPluginClasspath() ++ lineNumberPluginClasspath()
+
+    def lineNumberPluginClasspath: T[Agg[PathRef]] = T {
+      mill.modules.Util.millProjectModule(
+        "MILL_LINENUMBERS",
+        "mill-entrypoint-linenumbers",
+        repositoriesTask()
+      )
+    }
+  }
+}
+
+object MillBootstrapModule{
+  def top(relative: os.RelPath, base: os.Path, pkg: Seq[String], name: String) = {
+    val foreign =
+      if (pkg.nonEmpty || name != "build") {
+        // Computing a path in "out" that uniquely reflects the location
+        // of the foreign module relatively to the current build.
+
+        // Encoding the number of `/..`
+        val ups = if (relative.ups > 0) Seq(s"up-${relative.ups}") else Seq()
+        val segs =
+          Seq("foreign-modules") ++
+          ups ++
+          relative.segments.init ++
+          Seq(relative.segments.last.stripSuffix(".sc"))
+
+        val segsList = segs.map(pprint.Util.literalize(_)).mkString(", ")
+        s"Some(_root_.mill.define.Segments.labels($segsList))"
+      } else "None"
+    s"""
+       |package millbuild${pkg.map("." + _).mkString}
```

Comment:
We use `millbuild` as the package prefix for the `build.sc` contents and any script files it imports

Reply:
Is it necessary? Why can't we use the `empty` package?

Reply:
Good question. I know JVM empty packages are a bit unusual, so I chose to give it a name. Googling `java empty package` brings up enough question marks that I feel putting *some* package prefix is probably a good idea

Reply:
I was thinking that if we want in the future to migrate to vanilla Scala, people will need to add the package. But, thinking again, nothing in the build references the name, in that future, people will be able to use whatever package they want.

Reply:
I think, now that we use `ScalaModule` to compile, the file structure itself should not matter for performance. Zinc is tracking inter-dependencies.

Reply:
@lolgab The "empty package" which is called the "default package" in Java is special, as it can't be referred to from other packages. That means, you can declare classes in it and refer to other classes in it, but you can't refer to classes in the default package from classes in other packages or other jars. There are more details, but having some package name is probably the best convention for our scripts.